tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -8709,6 +8709,18 @@ SpirvEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
         callExpr, translateWaveOp(hlslOpcode, retType, srcLoc),
         spv::GroupOperation::ExclusiveScan);
   } break;
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixUSum:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixSum:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixUProduct:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixProduct:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixBitAnd:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixBitOr:
+  case hlsl::IntrinsicOp::IOP_WaveMultiPrefixBitXor: {
+    const auto retType = callExpr->getCallReturnType(astContext);
+    retVal = processWaveReductionOrPrefix(
+        callExpr, translateWaveOp(hlslOpcode, retType, srcLoc),
+        spv::GroupOperation::PartitionedExclusiveScanNV);
```

Comment:
Why is this using `PartitionedExclusiveScanNV` which is NV specific vs `ExclusiveScan` which is brought by the more generic `GroupNonUniformBallot`?
Seems like for exemple, WavePrefixSum could be implemented using `OpGroupNonUniformIAdd` and the `ExclusiveScan` group operation or am I missing something?

Reply:
The ballot (mask) operand is added as an operand to GroupNonUniform arithmetic instructions by the `SPV_NV_shader_subgroup_partitioned` extension, and then `PartitionedExclusiveScanNV` is the group operation that makes that operand required and and allows us to specify an arbitrary subset of the subgroup. Without it, you're doing an exclusive scan on the whole subgroup which is what the non-"Multi" WavePrefix HLSL intrinsics do.

Reply:
Ahh, for some reason I only looked at the WavePrefixSum part...
Thanks for the details! Yes, agree, don't see a way to represent this type of decoupage in the core part of the spec.